src/common/CommonStructsAndConstants.h

```diff
@@ -0,0 +1,229 @@
+     */
+    u_int32_t indexWriterVersion{0};
+
+    /**
+     * Keep track of the size of the stored index entries.
+     */
+    u_int32_t sizeOfIndexEntry{0};
+
+    /**
+     * A magic number to identify index files which were created
+     * with this software.
+     */
+    u_int32_t magicNumber = MAGIC_NUMBER;
+
+    /**
+     * The interval of blocks between index entries.
+     */
+    u_int32_t blockInterval{0};
+
+    /**
+     * This value is not available when a header is created in write mode. It will be written at the end of indexing
+     * process. However, the indexer and the extractor will work without it.
+     */
+    u_int64_t numberOfEntries{0};
+
+    /**
+     * Stores the amount of lines in the indexed file. This value is also written after the indexing is done.
+     * However, the indexer and the extractor will work without it.
+     */
+    u_int64_t linesInIndexedFile{0};
+
+    /**
+     * Tell the IndexReader, if the entries are compressed. (1-byte padded to 8 bytes!) The value looks weird without
+     * the followup value when you look it up in written fqi files. C++ does not do a clean write in this case.
+     */
+    bool dictionariesAreCompressed{false};
+    Bytef placeholder[7]{0};
+
+    /**
+     * Reserved space for information which might be added in
+     * the future.
+     */
+    u_int64_t reserved[59]{0};
+
+    explicit IndexHeader(u_int32_t binaryVersion, u_int32_t sizeOfIndexEntry, u_int32_t blockInterval, bool dictionariesAreCompressed) {
+        this->indexWriterVersion = binaryVersion;
+        this->sizeOfIndexEntry = sizeOfIndexEntry;
+        this->blockInterval = blockInterval;
+        this->dictionariesAreCompressed = dictionariesAreCompressed;
+    }
+
+    IndexHeader() = default;
+
+    bool operator==(const IndexHeader &rhs) const;
+
+    bool operator!=(const IndexHeader &rhs) const {
+        return !(rhs == *this);
+    }
+
+    explicit operator bool() const;
+
+    bool operator!() { return !this; }
+};
+
+/**
+ * A (decompressed / processed) representation of an IndexEntryV[n].
+ */
+struct IndexEntry {
+    u_int64_t id{0};
+    u_int64_t offsetInRawFile{0};
+    u_int64_t startingLineInEntry{0};
+    u_int64_t compressedDictionarySize{0};
+    u_int32_t bits{0};
+    u_int16_t offsetOfFirstValidLine{0};
+    Bytef window[WINDOW_SIZE]{0};
+
+    IndexEntry(u_int64_t id,
+               u_int32_t bits,
+               u_int16_t offsetOfFirstValidLine,
+               u_int64_t offsetInRawFile,
+               u_int64_t startingLineInEntry);
+
+    explicit IndexEntry(u_int64_t id) {
+        this->id = id;
+    }
+
+    IndexEntry() = default;
+
+    bool operator==(const IndexEntry &rhs) const;
+
+    bool operator!=(const IndexEntry &rhs) const { return !(rhs == *this); }
+
+};
+
+struct VirtualIndexEntry {
+    bool operator==(const VirtualIndexEntry &rhs) const { return true; };
+};
+
+/**
+ * Represents an entry in a gz index file
```

Comment:
Do you think for explanation would it be helpful to have e.g. a plant-UML diagram of the the blocks, and all other concepts related to gzip format and indexes and thus explain all the terms? Maybe we can make a sketch that I can translate into digital diagram (or we just scan the sketch!)